python/golem_task_api/testutils/dockertaskapiservice.py

```diff
@@ -0,0 +1,129 @@
+import os
+import subprocess
+import sys
+from pathlib import Path
+from typing import Tuple, Optional
+
+import docker
+from golem_task_api import constants, TaskApiService
+
+
+def is_docker_available():
+    """ Check the connection with Docker daemon """
+    try:
+        client = docker.from_env()
+        client.ping()
+    except Exception:
+        return False
+    return True
+
+
+def run_docker_container(
+        image: str,
+        work_dir: Path,
+        command: str,
+        port: int
+) -> docker.models.containers.Container:
+    """" Start a detached container """
+    if sys.platform == 'linux':
+        ports = {}
+    else:
+        ports = {port: port}
+
+    client = docker.from_env()
+    return client.containers.run(
+        image=image,
+        command=command,
+        volumes={
+            str(work_dir): {
+                'bind': f'/{constants.WORK_DIR}',
+                'mode': 'rw',
+            }
+        },
+        user=os.getuid(),
+        ports=ports,
+        detach=True,
+    )
+
+
+def get_docker_container_port_mapping(
+        container_id: str,
+        port: int,
+        vm_name: str,
+) -> Tuple[str, int]:
+    """ Retrieve running container's IP address and port """
+
+    client = docker.APIClient()
+    container_config = client.inspect_container(container_id)
+    net_config = container_config['NetworkSettings']
+
+    if sys.platform == 'darwin':
+        ip_address = '127.0.0.1'
+    elif sys.platform == 'win32':
+        ip_address = subprocess.check_output(['docker-machine', 'ip', vm_name])
+        ip_address = ip_address.decode('utf-8').strip()
+    else:
+        ip_address = net_config['Networks']['bridge']['IPAddress']
+
+    if sys.platform != 'linux':
+        port = int(net_config['Ports'][f'{port}/tcp'][0]['HostPort'])
+
+    if not ip_address:
+        raise RuntimeError(f'Unable to read the IP address of {container_id}')
+
+    return ip_address, port
+
+
+class DockerTaskApiService(TaskApiService):
+
+    def __init__(
+            self,
+            image: str,
+            work_dir: Path,
+            vm_name: Optional[str] = 'default'
+    ) -> None:
+        self._image = image
+        self._work_dir = work_dir
+        self._vm_name = vm_name
+        self._container = None
+
+    async def start(
+            self,
+            command: str,
+            port: int
+    ) -> Tuple[str, int]:
+        self._container = run_docker_container(
```

Comment:
The problem with this approach is that it uses synchronous calls which kind of defeats the purpose of making `start` method async. Maybe we could use [aiodocker](https://github.com/aio-libs/aiodocker) here?

Reply:
How did the original DockerTaskAPIService implementation got accepted then? This is based off of the original one and conforms to the service interface.

I don't think that aiodocker (https://github.com/golemfactory/task-api/pull/60 + working branch in tutorialapp repo) brings any added value to testutils. It requires changes to the base service interface and lacks in functionality.